### PR TITLE
Rouchon1 normalisation

### PR DESCRIPTION
**Pull request by pierreguilmin**
Replaces the work of @gautierronan from #249, mostly minor modifications and reorganisation. See description there.

### CODE REVIEW

File: dynamiqs/mesolve/rouchon.py [hunk 1]
```diff
@@ -104,21 +114,32 @@ def backward_augmented(
...
         M0 = self.M0(Hnh)
-        M0dag = self.M0dag(M0)
         M0rev = self.M0rev(Hnh)
 
+        # normalize the Kraus Map
+        if self.cholesky_normalization:
+            # Rrev, Trev: (b_H, 1, n, n)
+            Rrev = self.Rrev(M0rev)
+            Trev = self.Trev(Rrev)
+            rho = inv_mm_invmh(Trev.mH, rho)  # Trev.mH^-1 @ rho @ Trev^-1
```
Comment (gautierronan) [inline dynamiqs/mesolve/rouchon.py:128]: `Trev` is upper triangular, so `Trev.mH` is lower, and this doesn't work with your function `inv_mm_invmh`.

File: dynamiqs/mesolve/rouchon.py [hunk 2]
```diff
@@ -104,21 +114,32 @@ def backward_augmented(
...
         # compute phi(t-dt)
-        phi = kraus_map(phi, M0dag) + kraus_map(phi, self.M1sdag)
+        phi = kraus_map(phi, M0.mH) + kraus_map(phi, self.M1s.mH)
 
-        return rho, phi
+        if self.cholesky_normalization:
+            # R, T: (b_H, 1, n, n)
+            R = self.R(M0)
+            T = self.T(R)
+            phi = inv_mm_invmh(T, phi)  # T^-1 @ phi @ T.mH^-1
```
Comment (gautierronan) [inline dynamiqs/mesolve/rouchon.py:140]: Same here. `T` is lower, and `inv_mm_invmh` takes an upper triangular matrix.

File: dynamiqs/mesolve/rouchon.py [hunk 3]
```diff
@@ -80,11 +84,17 @@ def forward(self, t: float, rho: Tensor) -> Tensor:
         Hnh = self.Hnh(H)
         M0 = self.M0(Hnh)
 
+        # normalize the Kraus Map
+        if self.cholesky_normalization:
+            # R, T: (b_H, 1, n, n)
+            R = self.R(M0)
+            T = self.T(R)
```
Comment (gautierronan) [inline dynamiqs/mesolve/rouchon.py:91]: These two lines require caching two objects, whereas only `T` is used in the end. Not sure how important this is, I'll let you decide whether its worth correcting.

File: dynamiqs/mesolve/rouchon.py [hunk 4]
```diff
@@ -2,20 +2,24 @@
...
-    cache,
-    inv_sqrtm,
-    kraus_map,
-)
+from ..solvers.utils import cache, inv_sqrtm, kraus_map
 from ..utils.utils import unit
 from .me_solver import MESolver
 
 
+def inv_mm_invmh(A: Tensor, B: Tensor) -> Tensor:
```
Comment (gautierronan) [inline dynamiqs/mesolve/rouchon.py:15]: Probably a simple fix is to add the option `upper=True` in this function.

Comment (gautierronan) [reply dynamiqs/mesolve/rouchon.py:15]: How about `inv_kraus_matmul` for the name of this function? Not sure if it's better. I'll let you decide.

**Review by gautierronan — CHANGES_REQUESTED** [review]

File: dynamiqs/mesolve/rouchon.py (hunk 3 above)
Comment (pierreguilmin) [reply dynamiqs/mesolve/rouchon.py:91]: It's not 100% ideal but I'd prefer to keep it like that for now, it makes the code clearer, and we're also going to need this `R` matrix for the SME, the solvers will look more similar. I think we can refine performance later.

File: dynamiqs/mesolve/rouchon.py [hunk 5]
```diff
@@ -104,21 +113,32 @@ def backward_augmented(
...
+        if self.cholesky_normalization:
+            # Rrev, Trev: (b_H, 1, n, n)
+            Rrev = self.Rrev(M0rev)
+            Trev = self.Trev(Rrev)
+            rho = inv_kraus_matmul(Trev.mH, rho, upper=True)  # Tr.mH^-1 @ rho @ Tr^-1
+
         # compute rho(t-dt)
-        rho = kraus_map(rho, M0rev) - kraus_map(rho, self.M1srev)
-        rho = unit(rho)
+        rho = kraus_map(rho, M0rev) - kraus_map(rho, self.M1s)
```
Comment (gautierronan) [inline dynamiqs/mesolve/rouchon.py:130]:
```suggestion
        rho = kraus_map(rho, M0rev) - kraus_map(rho, self.M1srev)
```

File: dynamiqs/mesolve/rouchon.py [hunk 6]
```diff
@@ -24,43 +27,43 @@ class MERouchon1(MERouchon):
...
+                Hnh = self.Hnh(H)
+                M0 = self.M0(Hnh)
+                M0rev = self.M0rev(Hnh)
+
+                S = inv_sqrtm(self.R(M0))
+                Srev = inv_sqrtm(self.Rrev(M0rev))
+
+                self.M0 = cache(lambda Hnh: M0 @ S)
+                self.M1s = self.M1s @ S
+                self.M0rev = cache(lambda Hnh: M0rev @ Srev)
```
Comment (gautierronan) [inline dynamiqs/mesolve/rouchon.py:59]: Also, it should be
```python
self.M0rev = cache(lambda Hnh: Srev @ M0rev)
self.M1srev = Srev @ self.M1s
```
I think in order to preserve the unit Kraus map structure.

File: dynamiqs/mesolve/rouchon.py [hunk 7]
```diff
@@ -24,43 +27,43 @@ class MERouchon1(MERouchon):
...
+        # forward time operators
+        # M0: (b_H, 1, n, n)
+        # M1s: (1, len(L), n, n)
+        self.M0 = cache(lambda Hnh: self.I - 1j * self.dt * Hnh)
+        self.M1s = sqrt(self.dt) * self.L
+
+        # reverse time operators
+        # M0rev: (b_H, 1, n, n)
+        self.M0rev = cache(lambda Hnh: self.I + 1j * self.dt * Hnh)
+
```
Comment (gautierronan) [inline dynamiqs/mesolve/rouchon.py:39]:
```suggestion
self.M1srev = self.M1s
```

**Review by gautierronan — CHANGES_REQUESTED** [review]

**Comment by gautierronan** [comment]
After the corrections above, we can merge I think.

**Comment by pierreguilmin** [comment]
Alright, I'll just run some quick tests for time-dep Hamiltonian to make sure it works.

**Comment by pierreguilmin** [comment]
And then I'll let you rewrite the implem for the Rouchon2 solver on top.

**Comment by gautierronan** [comment]
> Alright, I'll just run some quick tests for time-dep Hamiltonian to make sure it works.

Yes I actually wrote down that we need tests for time-dependent problems. It's a missing piece in the library right now, also useful for the adjoint dopri PR.

**Review by gautierronan — APPROVED** [review]

**Comment by pierreguilmin** [comment]
> Yes I actually wrote down that we need tests for time-dependent problems. It's a missing piece in the library right now, also useful for the adjoint dopri PR.

Cf. task 27 of the [private roadmap](https://github.com/orgs/dynamiqs/projects/2). 😉

**Pull request merged**